<?xml version = "1.0"?>
<robot name = "handIA">

    <material name="grey">
        <color rgba="0.9 0.9 0.9 1"/>
    </material>

    <!-- <gazebo>
        <plugin name="ros_control" filename="libgazebo_ros2_control.so"/>
    </gazebo> -->

    <gazebo>
        <plugin filename="gz-sim-joint-state-publisher-system"
        name="gz::sim::systems::JointStatePublisher">
        <topic>joint_states</topic>
        <joint_name>blade_to_arm</joint_name>
        <joint_name>f1_to_hand</joint_name>
        <joint_name>f2_to_hand</joint_name>
        <joint_name>f3_to_hand</joint_name>
        <joint_name>f4_to_hand</joint_name>
        <joint_name>f5_to_hand</joint_name>
        </plugin>
        <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
            <joint_name>f1_to_hand</joint_name>
            <topic>f1</topic>
            <p_gain>1</p_gain>
            <i_gain>0.1</i_gain>
            <d_gain>0.01</d_gain>
            <i_max>1</i_max>
            <i_min>-1</i_min>
            <cmd_max>1000</cmd_max>
            <cmd_min>-1000</cmd_min>
        </plugin>
        <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
            <joint_name>f2_to_hand</joint_name>
            <topic>f2</topic>
            <p_gain>1</p_gain>
            <i_gain>0.1</i_gain>
            <d_gain>0.01</d_gain>
            <i_max>1</i_max>
            <i_min>-1</i_min>
            <cmd_max>1000</cmd_max>
            <cmd_min>-1000</cmd_min>
        </plugin>
        <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
            <joint_name>f3_to_hand</joint_name>
            <topic>f3</topic>
            <p_gain>1</p_gain>
            <i_gain>0.1</i_gain>
            <d_gain>0.01</d_gain>
            <i_max>1</i_max>
            <i_min>-1</i_min>
            <cmd_max>1000</cmd_max>
            <cmd_min>-1000</cmd_min>
        </plugin>
        <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
            <joint_name>f4_to_hand</joint_name>
            <topic>f4</topic>
            <p_gain>1</p_gain>
            <i_gain>0.1</i_gain>
            <d_gain>0.01</d_gain>
            <i_max>1</i_max>
            <i_min>-1</i_min>
            <cmd_max>1000</cmd_max>
            <cmd_min>-1000</cmd_min>
        </plugin>
        <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
            <joint_name>f5_to_hand</joint_name>
            <topic>f5</topic>
            <p_gain>1</p_gain>
            <i_gain>0.1</i_gain>
            <d_gain>0.01</d_gain>
            <i_max>1</i_max>
            <i_min>-1</i_min>
            <cmd_max>3</cmd_max>
            <cmd_min>-3</cmd_min>
        </plugin>

    </gazebo>


    <!-- //////////////////  arm   //////////////////-->
    <link name = "arm">
        <visual>
            <material name="grey"/>
            <geometry>
                <mesh filename = "package://roshandia_pkg/urdf/mesh/arm.stl"/>
            </geometry>
        </visual>
        <collision>
            <geometry>
                <mesh filename = "package://roshandia_pkg/urdf/mesh/arm.stl"/>
              </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>
    <!-- //////////////////  hand   //////////////////-->
    <link name ="hand">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/hand.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/hand.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name = "hand_to_arm" type="fixed">
        <parent link="arm"/>
        <child link="hand"/>
        <origin xyz="0 0 4.01612"/>
    </joint>
    <!-- //////////////////  blade   //////////////////-->
    <link name = "blade">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/blade.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/blade.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="blade_to_arm" type="prismatic">
        <parent link="arm"/>
        <child link="blade"/>
        <origin xyz="-0.836728 0.010775 1.07994"/>
        <axis xyz="0 0 1"/>
        <limit effort="100" velocity="0.2" upper="2.7037" lower="0"/>
    </joint>
    <!-- //////////////////  finger 1   //////////////////-->
    <link name = "finger1">
        <visual>
            <material name="grey"/>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger1.stl"/>
            </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger1.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="f1_to_hand" type="revolute">
        <parent link="hand"/>
        <child link="finger1"/>
        <origin xyz="0.001181 0.472818 3.3364"/>
        <axis xyz="0.707 0.707 0.707"/>
        <limit effort="100" velocity="2" lower="-1" upper="1"/>
    </joint>
    
    <!-- //////////////////  finger 2   //////////////////-->
    <link name = "finger2">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger2.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger2.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="f2_to_hand" type="revolute">
        <parent link="hand"/>
        <child link="finger2"/>
        <origin xyz="0.285932 0.615642 4.82351"/>
        <axis xyz="0 1 0"/>
        <limit effort="100" velocity="2" lower="-1.222" upper="0"/>
    </joint>
    <!-- //////////////////  finger 3   //////////////////-->
    <link name = "finger3">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger3.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger3.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="f3_to_hand" type="revolute">
        <parent link="hand"/>
        <child link="finger3"/>
        <origin xyz="0.260113 0.054977 4.73802"/>
        <axis xyz="0 1 0"/>
        <limit effort="100" velocity="2" lower="-1.222" upper="0"/>
    </joint>
    <!-- //////////////////  finger 4   //////////////////-->
    <link name = "finger4">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger4.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger4.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="f4_to_hand" type="revolute">
        <parent link="hand"/>
        <child link="finger4"/>
        <origin xyz="0.2049 -0.488003 4.61141"/>
        <axis xyz="0 1 0"/>
        <limit effort="100" velocity="0.5" lower="-1.222" upper="0"/>
    </joint>
    <!-- //////////////////  finger 5   //////////////////-->
    <link name = "finger5">
        <visual>
            <material name="grey"/>
             <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger5.stl"/>
             </geometry>
        </visual>
        <collision>
            <geometry> 
                <mesh filename = "package://roshandia_pkg/urdf/mesh/finger5.stl"/>
             </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <inertia
                ixx="0.001" ixy="0.0" ixz="0.0"
                iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
    </link>

    <joint name="f5_to_hand" type="revolute">
        <parent link="hand"/>
        <child link="finger5"/>
        <origin xyz="0.046572 -0.898537 4.52963"/>
        <axis xyz="0 1 0"/>
        <limit effort="100" velocity="0.5" lower="-1.222" upper="0"/>
    </joint>

<gazebo>
    <joint name='World' type='fixed'>
        <child>arm</child>
        <parent>world</parent>
        <axis>
          <xyz>0 0 1</xyz>
          <limit>
            <lower>-inf</lower>
            <upper>inf</upper>
            <effort>inf</effort>
            <velocity>inf</velocity>
            <stiffness>100000000</stiffness>
            <dissipation>1</dissipation>
          </limit>
          <dynamics>
            <spring_reference>0</spring_reference>
            <spring_stiffness>0</spring_stiffness>
            <damping>0</damping>
            <friction>0</friction>
          </dynamics>
        </axis>
        <pose>0 0 0 0 0 0</pose>
      </joint>
    </gazebo>
</robot>
